### ISSUE TITLE
`aws_security_group_rule` fails to create or destroy nondeterministically

### ISSUE THREAD

**Issue by bitglue** [hashicorp/terraform#3691]
Let's say I have an ELB, and some WWW servers, and a security group for each. And I'm using Terraform 0.6.6 to manage those security groups.

The ELB accepts inbound connections from anywhere, and can make outbound connections to the WWW servers.

The WWW servers accept inbound connections from the ELB.

That yields the Terraform configuration:

```
var "vpc_id" {}

resource "aws_security_group" "www_elb" {
    name = "www_elb"
    description = "load balancer"
    vpc_id = "${var.vpc_id}"
}   

# The ELB security group's rules must be separate resources, not inline, to
# avoid a dependency cycle between the ELB and the WWW security groups.

resource "aws_security_group_rule" "elb_http_ingress" {
    security_group_id = "${aws_security_group.www_elb.id}"
    type = "ingress"
    from_port = 80
    to_port = 80
    protocol = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
}   

resource "aws_security_group_rule" "elb_https_ingress" {
    security_group_id = "${aws_security_group.www_elb.id}"
    type = "ingress"
    from_port = 443
    to_port = 443
    protocol = "tcp"
    cidr_blocks = ["0.0.0.0/0"]
}   

resource "aws_security_group_rule" "elb_http_egress" {
    security_group_id = "${aws_security_group.www_elb.id}"
    type = "egress"
    from_port = 80
    to_port = 80
    protocol = "tcp"
    source_security_group_id = "${aws_security_group.www.id}"
}   

resource "aws_security_group_rule" "elb_https_egress" {
    security_group_id = "${aws_security_group.www_elb.id}"
    type = "egress"
    from_port = 443
    to_port = 443
    protocol = "tcp"
    source_security_group_id = "${aws_security_group.www.id}"
}   

resource "aws_security_group" "www" {
    name = "www"
    description = "web servers"
    vpc_id = "${var.vpc_id}"

    ingress {
        from_port = 22
        to_port = 22 
        protocol = "tcp"
        cidr_blocks = ["0.0.0.0/0"]
    }

    ingress {
        from_port = 80
        to_port = 80 
        protocol = "tcp"
        security_groups = ["${aws_security_group.www_elb.id}"]
    }

    ingress {
        from_port = 443
        to_port = 443
        protocol = "tcp"
        security_groups = ["${aws_security_group.www_elb.id}"]
    }

    egress {
        from_port = 0
        to_port = 0
        protocol = "-1"
        cidr_blocks = ["0.0.0.0/0"]
    }
}
```

Applying and deleting this configuration repeatedly results in failures, either on apply, or delete, or both.

Sometimes the apply fails like this:

```console
$ terraform apply
aws_security_group.www_elb: Creating...
aws_security_group.www_elb: Creation complete
aws_security_group_rule.elb_http_ingress: Creating...
aws_security_group_rule.elb_https_ingress: Creating...
aws_security_group.www: Creating...
aws_security_group_rule.elb_https_ingress: Creation complete
aws_security_group.www: Creation complete
aws_security_group_rule.elb_http_egress: Creating...
aws_security_group_rule.elb_https_egress: Creating...
aws_security_group_rule.elb_https_egress: Creation complete
aws_security_group_rule.elb_http_egress: Creation complete
Error applying plan:

1 error(s) occurred:

* aws_security_group_rule.elb_http_ingress: [WARN] No ingress rules were found for Security Group (www_elb) looking for Security Group Rule (sgrule-5174297)
```

On this run, a subsequent `terraform plan` showed no changes and everything looked to be created fine.

On another run I got a slightly different error:

```
Error applying plan:

2 error(s) occurred:

* aws_security_group_rule.elb_https_egress: [WARN] No egress rules were found for Security Group (www_elb) looking for Security Group Rule (sgrule-817536930)
* aws_security_group_rule.elb_http_egress: [WARN] No egress rules were found for Security Group (www_elb) looking for Security Group Rule (sgrule-4276344808)
```

Any subsequent `terraform plan` then yields the same error message. The only way to recover from this situation seems to be `terraform destroy -refresh=false`, or I suppose some manual mucking in the state file.

Separately, `terraform destroy` will also sometimes fail, and best I can tell, it's because one of the egress rules somehow disappears from the state file!

```console
$ terraform apply
aws_security_group.www_elb: Creating...
  description: "" => "load balancer"
  egress.#:    "" => "<computed>"
  ingress.#:   "" => "<computed>"
  name:        "" => "www_elb"
  owner_id:    "" => "<computed>"
  vpc_id:      "" => "vpc-7ac6701e"
aws_security_group.www_elb: Creation complete
aws_security_group_rule.elb_https_ingress: Creating...
  cidr_blocks.#:            "" => "1"
  cidr_blocks.0:            "" => "0.0.0.0/0"
  from_port:                "" => "443"
  protocol:                 "" => "tcp"
  security_group_id:        "" => "sg-3d339b5b"
  self:                     "" => "0"
  source_security_group_id: "" => "<computed>"
  to_port:                  "" => "443"
  type:                     "" => "ingress"
aws_security_group_rule.elb_http_ingress: Creating...
  cidr_blocks.#:            "" => "1"
  cidr_blocks.0:            "" => "0.0.0.0/0"
  from_port:                "" => "80"
  protocol:                 "" => "tcp"
  security_group_id:        "" => "sg-3d339b5b"
  self:                     "" => "0"
  source_security_group_id: "" => "<computed>"
  to_port:                  "" => "80"
  type:                     "" => "ingress"
aws_security_group.www: Creating...
  description:                                   "" => "web servers"
  egress.#:                                      "" => "1"
  egress.482069346.cidr_blocks.#:                "" => "1"
  egress.482069346.cidr_blocks.0:                "" => "0.0.0.0/0"
  egress.482069346.from_port:                    "" => "0"
  egress.482069346.protocol:                     "" => "-1"
  egress.482069346.security_groups.#:            "" => "0"
  egress.482069346.self:                         "" => "0"
  egress.482069346.to_port:                      "" => "0"
  ingress.#:                                     "" => "3"
  ingress.1428579889.cidr_blocks.#:              "" => "0"
  ingress.1428579889.from_port:                  "" => "80"
  ingress.1428579889.protocol:                   "" => "tcp"
  ingress.1428579889.security_groups.#:          "" => "1"
  ingress.1428579889.security_groups.2615877241: "" => "sg-3d339b5b"
  ingress.1428579889.self:                       "" => "0"
  ingress.1428579889.to_port:                    "" => "80"
  ingress.2541437006.cidr_blocks.#:              "" => "1"
  ingress.2541437006.cidr_blocks.0:              "" => "0.0.0.0/0"
  ingress.2541437006.from_port:                  "" => "22"
  ingress.2541437006.protocol:                   "" => "tcp"
  ingress.2541437006.security_groups.#:          "" => "0"
  ingress.2541437006.self:                       "" => "0"
  ingress.2541437006.to_port:                    "" => "22"
  ingress.3106363127.cidr_blocks.#:              "" => "0"
  ingress.3106363127.from_port:                  "" => "443"
  ingress.3106363127.protocol:                   "" => "tcp"
  ingress.3106363127.security_groups.#:          "" => "1"
  ingress.3106363127.security_groups.2615877241: "" => "sg-3d339b5b"
  ingress.3106363127.self:                       "" => "0"
  ingress.3106363127.to_port:                    "" => "443"
  name:                                          "" => "www"
  owner_id:                                      "" => "<computed>"
  vpc_id:                                        "" => "vpc-7ac6701e"
aws_security_group_rule.elb_https_ingress: Creation complete
aws_security_group_rule.elb_http_ingress: Creation complete
aws_security_group.www: Creation complete
aws_security_group_rule.elb_https_egress: Creating...
  from_port:                "" => "443"
  protocol:                 "" => "tcp"
  security_group_id:        "" => "sg-3d339b5b"
  self:                     "" => "0"
  source_security_group_id: "" => "sg-39339b5f"
  to_port:                  "" => "443"
  type:                     "" => "egress"
aws_security_group_rule.elb_http_egress: Creating...
  from_port:                "" => "80"
  protocol:                 "" => "tcp"
  security_group_id:        "" => "sg-3d339b5b"
  self:                     "" => "0"
  source_security_group_id: "" => "sg-39339b5f"
  to_port:                  "" => "443"
  type:                     "" => "egress"
aws_security_group_rule.elb_http_egress: Creating...
  from_port:                "" => "80"
  protocol:                 "" => "tcp"
  security_group_id:        "" => "sg-3d339b5b"
  self:                     "" => "0"
  source_security_group_id: "" => "sg-39339b5f"
  to_port:                  "" => "80"
  type:                     "" => "egress"
aws_security_group_rule.elb_https_egress: Creation complete
aws_security_group_rule.elb_http_egress: Creation complete

Apply complete! Resources: 6 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate
$ terraform destroy -force
aws_security_group.www_elb: Refreshing state... (ID: sg-3d339b5b)
aws_security_group.www: Refreshing state... (ID: sg-39339b5f)
aws_security_group_rule.elb_http_ingress: Refreshing state... (ID: sgrule-609903174)
aws_security_group_rule.elb_https_ingress: Refreshing state... (ID: sgrule-2747051615)
aws_security_group_rule.elb_https_egress: Refreshing state... (ID: sgrule-3299350019)
aws_security_group_rule.elb_https_egress: Destroying...
aws_security_group_rule.elb_http_ingress: Destroying...
aws_security_group_rule.elb_https_ingress: Destroying...
aws_security_group_rule.elb_http_ingress: Destruction complete
aws_security_group_rule.elb_https_egress: Destruction complete
aws_security_group.www: Destroying...
aws_security_group_rule.elb_https_ingress: Destruction complete
Error applying plan:

1 error(s) occurred:

* aws_security_group.www: DependencyViolation: resource sg-39339b5f has a dependent object
        status code: 400, request id: 
```

Terraform can't destroy aws_security_group.www because it hasn't destroyed aws_security_group_rule.elb_http_egress, so there's still a rule on the ELB group referencing the WWW group. And after this failure, I don't see it in the state file. But reading the output of `terraform apply`, it was clearly created! :confused:

**Comment by packplusplus** [comment]
@clstokes this is the same as we talked about right?

**Comment by mrwilby** [comment]
+1, also just hit this

**Comment by packplusplus** [comment]
@mrwilby I played with the latest master and it appears the be fixed. Not advocating you run master, just that the next release will probably fix this issue.

**Comment by mrwilby** [comment]
@packplusplus Thanks, good to know!

**Comment by catsby** [comment]
Hey all – thanks for writing in. This should be fixed with https://github.com/hashicorp/terraform/pull/3965 , which is included in the new release v0.6.7 that I just shipped. 

Thanks! Closing this for now, please let me know if you're still hitting this
